crates/supervisor/storage/src/providers/log_provider.rs

```diff
@@ -0,0 +1,383 @@
+//! Reth's MDBX-backed abstraction of [`LogProvider`] for superchain state.
+//!
+//! This module provides the [`LogProvider`] struct, which uses the
+//! [`reth-db`] abstraction of reth to store execution logs
+//! and block metadata required by the Optimism supervisor.
+//!
+//! It supports:
+//! - Writing full blocks of logs with metadata
+//! - Retrieving block metadata by number
+//! - Finding a block from a specific log (with hash/index match)
+//! - Fetching logs per block using dup-sorted key layout
+//!
+//! Logs are stored in [`LogEntries`] under dup-sorted tables, with log index
+//! used as the subkey. Block metadata is stored in [`BlockRefs`].
+
+use crate::{
+    error::StorageError,
+    models::{BlockRefs, LogEntries},
+};
+use kona_protocol::BlockInfo;
+use kona_supervisor_types::Log;
+use reth_db_api::{
+    cursor::{DbCursorRO, DbDupCursorRO, DbDupCursorRW},
+    transaction::{DbTx, DbTxMut},
+};
+use std::fmt::Debug;
+use tracing::{debug, error, warn};
+
+/// A log storage that wraps a transactional reference to the MDBX backend.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct LogProvider<'tx, TX> {
+    tx: &'tx TX,
+}
+
+/// Internal constructor and setup methods for [`LogProvider`].
+impl<'tx, TX> LogProvider<'tx, TX> {
+    #[allow(dead_code)]
+    pub(crate) const fn new(tx: &'tx TX) -> Self {
+        Self { tx }
+    }
+}
+
+impl<TX> LogProvider<'_, TX>
+where
+    TX: DbTxMut + DbTx,
+{
+    fn store_block_logs(&self, block: &BlockInfo, logs: Vec<Log>) -> Result<(), StorageError> {
+        debug!(target: "supervisor_storage", block_number = block.number, "Storing logs");
+
+        if let Ok(latest_block) = self.get_latest_block() {
+            if latest_block.number + 1 != block.number || latest_block.hash != block.parent_hash {
+                warn!(
+                    target: "supervisor_storage",
+                    latest_block = ?latest_block,
+                    incoming_block = ?block,
+                    "Incoming block does not follow latest stored block"
+                );
+                return Err(StorageError::ConflictError(
+                    "Incoming block does not follow latest stored block".into(),
+                ));
+            }
+        }
```

Comment:
Shouldn't we check for genesis?

If we think that it will be decided later on who will be responsible for this check and we can add a todo `todo: analyze if we should check for genesis block`

Reply:
So, the current scenario is if there is no block in db it will insert what is being sent at first.
I am not sure everytime we need to start from genesis or not. Or we can also start from a specific height. That's why the check is not here in db level.

Reply:
Ok. I will check this at the chaindb issue.